In [32]:
# Importing modules
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
from sklearn.neighbors import KNeighborsClassifier
from sklearn.metrics import confusion_matrix
from sklearn.metrics import f1_score
from sklearn.metrics import accuracy_score

In [33]:
# read the CSV file
df = pd.read_csv('spambase.csv', header=None)
# Printing the length of the dataset
len(df)

4601

In [34]:
# Let's have a look at the dataset
df.head()

0     1     2    3     4     5     6     7     8     9   ...    48  \
0  0.00  0.64  0.64  0.0  0.32  0.00  0.00  0.00  0.00  0.00  ...  0.00   
1  0.21  0.28  0.50  0.0  0.14  0.28  0.21  0.07  0.00  0.94  ...  0.00   
2  0.06  0.00  0.71  0.0  1.23  0.19  0.19  0.12  0.64  0.25  ...  0.01   
3  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   
4  0.00  0.00  0.00  0.0  0.63  0.00  0.31  0.63  0.31  0.63  ...  0.00   

      49   50     51     52     53     54   55    56  57  
0  0.000  0.0  0.778  0.000  0.000  3.756   61   278   1  
1  0.132  0.0  0.372  0.180  0.048  5.114  101  1028   1  
2  0.143  0.0  0.276  0.184  0.010  9.821  485  2259   1  
3  0.137  0.0  0.137  0.000  0.000  3.537   40   191   1  
4  0.135  0.0  0.135  0.000  0.000  3.537   40   191   1  

[5 rows x 58 columns]

In [35]:
# Assign the column labels
df.columns = [
    'word_freq_make',
    'word_freq_address',
    'word_freq_all',
    'word_freq_3d',
    'word_freq_our',
    'word_freq_over',
    'word_freq_remove',
    'word_freq_internet',
    'word_freq_order',
    'word_freq_mail',
    'word_freq_receive',
    'word_freq_will',
    'word_freq_people',
    'word_freq_report',
    'word_freq_addresses',
    'word_freq_free',
    'word_freq_business',
    'word_freq_email',
    'word_freq_you',
    'word_freq_credit',
    'word_freq_your',
    'word_freq_font',
    'word_freq_000',
    'word_freq_money',
    'word_freq_hp',
    'word_freq_hpl',
    'word_freq_george',
    'word_freq_650',
    'word_freq_lab',
    'word_freq_labs',
    'word_freq_telnet',
    'word_freq_857',
    'word_freq_data',
    'word_freq_415',
    'word_freq_85',
    'word_freq_technology',
    'word_freq_1999',
    'word_freq_parts',
    'word_freq_pm',
    'word_freq_direct',
    'word_freq_cs',
    'word_freq_meeting',
    'word_freq_original',
    'word_freq_project',
    'word_freq_re',
    'word_freq_edu',
    'word_freq_table',
    'word_freq_conference',
    'char_freq_;',
    'char_freq_(',
    'char_freq_[',
    'char_freq_!',
    'char_freq_$',
    'char_freq_#',
    'capital_run_length_average',
    'capital_run_length_longest',
    'capital_run_length_total',
    'spam'
]

In [36]:
# Now let's look at the labelled data
df.sample(n = 100)

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
2241            0.00               0.00           0.00           0.0   
602             0.00               0.58           0.58           0.0   
2651            0.00               0.00           1.09           0.0   
2117            0.05               0.00           0.15           0.0   
1020            0.05               0.05           0.40           0.0   
...              ...                ...            ...           ...   
563             0.00               0.45           0.00           0.0   
2898            0.00               0.00           1.03           0.0   
4020            0.00               0.00           0.00           0.0   
4245            0.28               0.00           0.28           0.0   
4571            0.00               0.00           0.46           0.0   

      word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
2241           0.00            0.00              0.00                0.00   
602            0.00            0.00              0.00                0.58   
2651           0.00            0.00              0.00                0.00   
2117           0.05            0.05              0.00                0.00   
1020           0.34            0.00              0.00                0.00   
...             ...             ...               ...                 ...   
563            0.91            0.00              1.36                0.00   
2898           2.06            0.00              0.00                0.00   
4020           0.00            0.00              0.00                0.00   
4245           0.00            0.00              0.00                0.00   
4571           0.23            0.23              0.00                0.00   

      word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
2241             0.00            0.00  ...        0.000        0.058   
602              0.58            7.55  ...        0.000        0.192   
2651             0.00            0.00  ...        0.000        0.000   
2117             0.52            0.00  ...        0.223        0.162   
1020             0.57            0.05  ...        0.019        0.099   
...               ...             ...  ...          ...          ...   
563              0.00            0.00  ...        0.000        0.000   
2898             0.00            0.00  ...        0.373        0.186   
4020             0.00            0.00  ...        0.000        1.190   
4245             0.00            0.00  ...        0.000        0.103   
4571             0.00            0.00  ...        0.000        0.082   

      char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
2241        0.000        0.000        0.058        0.058   
602         0.000        0.192        0.000        0.096   
2651        0.000        0.000        0.000        0.000   
2117        0.084        0.000        0.015        0.000   
1020        0.000        0.099        0.079        0.009   
...           ...          ...          ...          ...   
563         0.000        0.368        0.000        0.000   
2898        0.000        0.000        0.000        0.000   
4020        0.000        0.000        0.000        0.000   
4245        0.000        5.054        0.000        0.000   
4571        0.000        0.082        0.000        0.000   

      capital_run_length_average  capital_run_length_longest  \
2241                       1.755                           9   
602                        1.526                          10   
2651                       1.111                           2   
2117                       2.725                          38   
1020                       4.906                          95   
...                          ...                         ...   
563                        1.680                          17   
2898                       1.529                           5   
4020                       3.333                      

In [37]:
# Setting the spam value to 0 if the "word_freq_george" or the "word_freq_650" columns are greater than 0.0
# Reason: In the documentation(spambase.DOCUMENTATION) it has clearly mentioned it
df.loc[(df['word_freq_george'] > 0) | (df['word_freq_650'] > 0), 'spam'] = 0

# Save the modified dataframe to a new csv file
df.to_csv('modified_dataset.csv', index=False)


In [42]:
# Creating a reference to the updated dataset
df_new = pd.read_csv('modified_dataset.csv', header=0)
df_new.head()

word_freq_make  word_freq_address  word_freq_all  word_freq_3d  \
0            0.00               0.64           0.64           0.0   
1            0.21               0.28           0.50           0.0   
2            0.06               0.00           0.71           0.0   
3            0.00               0.00           0.00           0.0   
4            0.00               0.00           0.00           0.0   

   word_freq_our  word_freq_over  word_freq_remove  word_freq_internet  \
0           0.32            0.00              0.00                0.00   
1           0.14            0.28              0.21                0.07   
2           1.23            0.19              0.19                0.12   
3           0.63            0.00              0.31                0.63   
4           0.63            0.00              0.31                0.63   

   word_freq_order  word_freq_mail  ...  char_freq_;  char_freq_(  \
0             0.00            0.00  ...         0.00        0.000   
1             0.00            0.94  ...         0.00        0.132   
2             0.64            0.25  ...         0.01        0.143   
3             0.31            0.63  ...         0.00        0.137   
4             0.31            0.63  ...         0.00        0.135   

   char_freq_[  char_freq_!  char_freq_$  char_freq_#  \
0          0.0        0.778        0.000        0.000   
1          0.0        0.372        0.180        0.048   
2          0.0        0.276        0.184        0.010   
3          0.0        0.137        0.000        0.000   
4          0.0        0.135        0.000        0.000   

   capital_run_length_average  capital_run_length_longest  \
0                       3.756                          61   
1                       5.114                         101   
2                       9.821                         485   
3                       3.537                          40   
4                       3.537                          40   

   capital_run_length_total  spam  
0                       278     1  
1                      1028     1  
2                      2259     1  
3                       191     1  
4                       191     1  

[5 rows x 58 columns]

In [43]:
# Let's have a look at the new dataset and check if its updated or not
columns_to_display = ['word_freq_george', 'word_freq_650', 'spam']
df_subset = df_new.loc[:, columns_to_display]
df_subset

word_freq_george  word_freq_650  spam
0                  0.0            0.0     1
1                  0.0            0.0     1
2                  0.0            0.0     1
3                  0.0            0.0     1
4                  0.0            0.0     1
...                ...            ...   ...
4596               0.0            0.0     0
4597               0.0            0.0     0
4598               0.0            0.0     0
4599               0.0            0.0     0
4600               0.0            0.0     0

[4601 rows x 3 columns]

In conclusion, we can state that if the values in the "word_freq_george" or "word_freq_650" columns are greater than 0.0, then the corresponding values in the "spam" column have been updated to 0. This implies that the occurrence of the words "george" or "650" in an email are strong indicators that the email is not spam.

In [44]:
# Splitting the data into test and train
X = df_new.iloc[:, 0:57]
y = df_new.iloc[:, 57]
X_train, X_test, y_train, y_test = train_test_split(X,y,random_state=0, test_size=0.2)

In [45]:
# Feature scaling
sc_X = StandardScaler()
X_train = sc_X.fit_transform(X_train)
X_test = sc_X.transform(X_test)

In [46]:
# Finding the value of n_neighbors
import math
math.sqrt(len(y_test))

30.347981810987037

Always the n_neighbors value has to be an odd number, therefore n_neighbors value is 29
n_neighbor = len(y_test) - 1

In [47]:
# Define the KNN model
# Setting p value into 2, because "spam" column has only 2 values, either 0 or 1
classifier = KNeighborsClassifier(n_neighbors=29, p=2, metric='euclidean')
classifier.fit(X_train, y_train)

KNeighborsClassifier(metric='euclidean', n_neighbors=29)

In [48]:
# Predict the test set result
y_pred = classifier.predict(X_test)
y_pred

array([1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 0, 0, 0,
       0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0, 0,
       1, 1, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0, 1,
       0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0,
       1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0,
       0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 1, 1, 0,
       1, 1, 1, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 0, 1, 1, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1, 1, 0, 0, 1, 0, 0, 1, 1,
       0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 0, 0, 1, 1, 0, 1, 0, 1, 0, 1, 1, 1,
       1, 1, 1, 1, 1, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 1, 0, 0, 0, 0, 0,
       0, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 0,
       0, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0, 0, 1, 0, 0, 0,
       0, 1, 0, 1, 0, 1, 1, 0, 0, 1, 1, 1, 1, 0, 0, 0, 1, 0, 0, 1, 1, 1,
       0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 0, 0,

In [49]:
# Evaluate the model using confusion matrix
cm = confusion_matrix(y_test, y_pred)
cm

array([[509,  38],
       [ 67, 307]], dtype=int64)

In the above matrix, true positive value is 307, which means that the model correctly identified 307 spam emails.
The true negative (TN) value is 509, which means that the model correctly identified 509 non-spam emails.
The false positive (FP) value is 38, which means that the model incorrectly identified 38 non-spam emails as spam.
The false negative (FN) value is 67, which means that the model incorrectly identified 67 spam emails as non-spam.

In [50]:
# printing the f1_score
f1_score(y_test, y_pred)

0.8539638386648123

In [51]:
# printing the accuracy_score
accuracy_score(y_test, y_pred)

0.8859934853420195